File: 118_Pascals_Triangle.cpp
/*
Given a non-negative integer numRows, 
generate the first numRows of Pascal's triangle.

Input: 5
Output:
[
     [1],
    [1,1],
   [1,2,1],
  [1,3,3,1],
 [1,4,6,4,1]
]
*/
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    
    public:
        vector<vector<int>> generate(int numRows) {
            if(numRows == 0) return {};
            vector<vector<int>> l;
            vector<int> last = {1};
            l.push_back(last);
            int n = 2;
            while(n <= numRows) {
                int i = 0;
                vector<int> pas = {1};
                for(i = 0; i < n - 2; i++) {
                    pas.push_back(last[i] + last[i+1]);
                }
                pas.push_back({1});
                l.push_back(pas);
                last = pas;
                n++;
            }
            return l;

        }
        
};

int main() {
    Solution sol;
    return 0;
}

File: py/118_Pascals_Triangle.py
from typing import List

class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        res, n = [], 1
        while n <= numRows:
            if n < 3:
                res.append([1 for i in range(n)])
            else:
                tmp = [1]
                last = res[n - 2]
                for i in range(n - 2):
                    tmp.append(last[i] + last[i + 1])
                tmp.append(1)
                res.append(tmp)
            n += 1
        return res
